Save multiple checkpoints in db in preparation to support rollbacks

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#643 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have adapted put/readCheckpoint and added unit tests
- [x] I have changed TxMeta to accomodate needed changes
- [x] I have adapted put/readTxHistory and added unit tests
- [x] I have supported and united random and sequential scheme tests


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
add very preliminary scaffolding of more detailed tests of sqlite spec

paweljakubas committed:
adapt putCheckpoint to multiple-checkpoints reality

update readCheckpoint accoring to decision

paweljakubas committed:
Change the primary key of TxMeta table

paweljakubas committed:
add further unit tests

post rebase corrections

fix http-bridge bench to compile

paweljakubas committed:
fix bugs and add incoming tx unit test

correct put/readCheckpoint and extend unit test

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -677,6 +685,17 @@ deleteCheckpoints wid = do
     deleteState @s wid -- clear state
 
+-- | Delete the checkpoint associated with a wallet and slotId.
+deleteCheckpoint
```
this one is introduced to have guarantee we can put the checkpoints with the same id more than 1 time

rvl replied:
Seems ok. Probably easier than changing the tests to not put the same checkpoint twice.

paweljakubas replied:
I enforced this behavior as I remember we wanted this kind of property from db. But it was when MVar version was implemented

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -756,8 +775,32 @@ selectUTxO (Checkpoint wid sl _parent _bh) = fmap entityVal <$>
+        , UtxoSlotSpent !=. Nothing
         ] []
 
+updateUTxOs
```
This one is introduced to change utxoslotspent of  all utxos in the last checkpoint to Nothing

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -90,7 +90,7 @@ TxMeta
     txMetaAmount     Natural      sql=amount
 
-    Primary txMetaTxId txMetaWalletId
+    Primary txMetaTxId txMetaWalletId txMetaSlotId
```
This one is to store both states for outcoming tx when we have Pending for slot0 and InLedger for slot1, where slot1 is later than slot0. Without it we cannot revert to Pending when rollback

paweljakubas commented on lib/http-bridge/test/bench/Cardano/Wallet/Primitive/AddressDiscovery/Any.hs:
```diff
@@ -95,9 +95,9 @@ instance PersistState AnyAddressState where
-
     deleteState wid =
         deleteWhere [ DB.AnyAddressStateWalletId ==. wid ]
+    deleteCheckpointState _wid _sl = undefined
```
I was not thinking alot here as it is in http-bridge

rvl replied:
This could be `pure ()`. This `Any` scheme code will be moved across to jormungandr in due course.

paweljakubas replied:
indeed, done

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -159,6 +191,126 @@ sqliteSpecRnd =
         describe "Sqlite State machine (RndState)" $ do
             it "Sequential state machine tests" prop_sequential
 
+runSqlite'
```
this stuff requires not-so-small number of dependencies, but I needed it to compile the stuff and use SQL queries side by side of db layer calls

rvl replied:
We need to figure out a way of easily switching on query logging for debugging tests. It should all be possible through iohk-monitoring.

paweljakubas replied:
I need `runSqlite'` mainly to use SQL SELECT - and I need db methods along with SQL queries just to precisely check that everything is as expected in db - I want to inspect number of UTxO rows, exact column values. Not how they are retrieved using db methods readCheckpoint

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -310,7 +314,11 @@ newDBLayer logConfig trace fp = do
+                      -- support updating of checkpoint
+                      let (BlockHeader sid _) = currentTip cp
+                      deleteCheckpoint @s wid sid
+                      updateUTxOs wid
```
`updateUTxOs` where utxo are soft-deleted should go to `putTxHistory` - will change it in the forthcomming commit

paweljakubas committed:
reshuffle tests and start outgoing tx unit test

paweljakubas committed:
add more tests and adapt putTxHistory

make code more reusable

use one row per utxo scheme

nix patch

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -159,6 +199,297 @@ sqliteSpecRnd =
+          let utxoEntry =
+                  TH.UTxO wid nextSlotId
+                  Nothing (TxId txid) ix addr amt
+          runSqlite' "spec.db" $ do
```
oh, it will be better to use `unsafeInitWallet` as used above here

paweljakubas replied:
done

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -756,8 +773,44 @@ selectUTxO (Checkpoint wid sl _parent _bh) = fmap entityVal <$>
+    :: [TxMeta]
+    -> [TxIn]
+    -> SqlPersistT IO ()
+addSpentUtxos metas txins = do
```
This needs to be an update not an insert. Because the utxo that is spent should already be there.

For an incoming transaction, the utxo entries are already getting added as part of the checkpoint.
For outgoing transactions, we want to soft-delete the utxo that has been spent, which can be done by setting slot_spent.

With transactions, we change from Pending -> InChain by adding new rows, but for utxo we update the existing rows.

paweljakubas replied:
@rvl ok. so let's say we have utxo0 at some slot0. if I understand well we want :
- new row with `(utxo0, slot0, Nothing)`
Then let's say slot incremented. now after putCheckpoint we want to have two rows:
```
(utxo0, slot0, Nothing)
(utxo0, slot1, Nothing)
```
Now we are sending tx (from the point of view of the wallet it is outgoing. So we are for example doing this in slot2 so we have 3 rows:
```
(utxo0, slot0, Nothing)
(utxo0, slot1, Nothing)
(utxo0, slot2, Nothing)
```
Now the block is coming and applied at slot3. if we have change then this wallet should see:
```
(utxo0, slot0, Nothing)
(utxo0, slot1, Nothing)
(utxo0, slot2, (Just slot3))
(utxo1,slot3,Nothing)
```
and not 
```
(utxo0, slot0, Nothing)
(utxo0, slot1, Nothing)
(utxo0, slot2, Nothing)
(utxo0, slot3, (Just slot3))
(utxo1,slot3,Nothing)
```
Right? I did intentionally the second version, but now the one above looks ok (especially if we rollback - here we will need to just erase rows). Or maybe you were contemplating the version that we have only one row per utxo which is only written twice : when appears after applying block and then when spent (here Nothing -> Just someSlot). When rollback then we will just turn back from (Just someslot) -> Nothing when rollback deeper when someslot or erase the row when slot0 is later than the depth of utxo

paweljakubas replied:
now thinking that maximum one row per utxo is better option. and here, indeed, update will be used

paweljakubas replied:
in the coming commit I will adopt this kind of behavior (excerpt from private communication with Rodney):
```
ok, so basically we have one row per utxo. it has this kind of lifetime:
(utxo0, slot0, Nothing) - slot0...slot2
(utxo0, slot0, (Just slot3)) for slot3 and later
```

paweljakubas replied:
one row per utxo logic depicted above is now implemented. all tests adopted

KtorZ replied:
>  Then let's say slot incremented. now after putCheckpoint we want to have two rows:

No, if we create a new checkpoint with the same UTxO, we do not duplicate the UTxO since nothing has changed. So, the UTxO is pretty much immutable except for the `slotSpent` column which is populated when we spend the UTxO :+1:

rvl reviewed (commented):
:+1: Getting there

rvl committed:
SqliteSpec: Add helper functions

rvl committed:
DB.Model: Add saving of checkpoints

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -664,7 +671,7 @@ insertCheckpoint wid cp = do
     putTxMetas metas
     putTxs ins outs
-    dbChunked insertMany_ utxo
+    mapM_ (insertWhenNew wid) utxo
```
You should be able to replace this with `mapM_ insertUniqueEntity utxo`. It will be faster than insertWhenNew. Hopefully the db bench will show us how much putting checkpoints slows down with the rollback changes.

KtorZ replied:
This change looks premature to me, we don't know yet if it's that of a problem to insert UTxO. Since we only insert a checkpoint every now and then ... I barely think this would be required. 

Plus, we are loosing a guard provided by `dbChunked` which covered our back regarding too big queries. Now we have re-introduced this problem :grimacing: 

Suggested fix: 

- Do not change this code now and wait to see if it's necessary
- Remove `insertWhenNew` function

paweljakubas replied:
if we do not change this block then we will have for the same UTxO many times as time passes, ie for slots slot0, slot1, slot2,... insertWhenNew makes sure we insert once only. Thanks to that we will have slot0..when putTxHistory we will change spent_utxo from Nothing to (just slotSpent) within the same row. If I understand well this is what we want, right?

KtorZ replied:
What happens in persistent if we try to insert an entity that already exists? Isn't there any sort of `tryInsert` where we can try to insert something, and silently ignore the query if the entity already exists.. ?

KtorZ replied:
:point_up: seems like `upsert` would work for this? It inserts if not exists, and we can simply have an identity as the udpate function if it exists.

paweljakubas replied:
Looking at UTxO table we have : 
```
158     Primary                                                                                                                                                                               
159         utxoWalletId                                                                                                                                                                      
160         utxoSlot                                                                                                                                                                          
161         utxoInputId                                                                                                                                                                       
162         utxoInputIndex                                                                                                                                                                    
163         utxoOutputAddress                                                                                                                                                                 
164         utxoOutputCoin
```
meaning if I have one utxo row with let's say `utxoSlot=slot1` , and do NOT want to insert when `utxoSlot=slot3` with other column values the same. So I need to check if the row for all column values beside utxoSlot is existent. If not then insert, otherwise leave it. But at the time slot3 I do not have information without select with proper filter that this same utxo was already inserted at `utxoSlot=slot1`

paweljakubas replied:
tried something like that : 
```
dbChunked insertWhenNew1 utxo

781  insertWhenNew1                                                                                                                                                                           
782      :: [UTxO]                                                                                                                                                                            
783      -> SqlPersistT IO ()                                                                                                                                                                 
784  insertWhenNew1 utxos = do                                                                                                                                                                
785      mapM_ insertUniqueEntity utxos
```
but failed;-/ more utxos were inserted

paweljakubas replied:
nevertheless, refactored the current version to be used with `dbChunked`

rvl replied:
`insertUniqueEntity` _is_ supposed to be the "try insert and ignore failure" function that we want -- at least according to my reading of the persistent docs. I will investigate.

rvl replied:
I have changed it to use `dbChunked putMany utxos`. It's the best option I could find in the persistent docs.

`dbChunked` is not necessary when using `mapM_` and inserting _single_ records.

But we want to use the batch insert operations if at all possible.

We need to handle the case of conflicting utxo because it is inserting the entire wallet utxo every checkpoint. It would be nice to be able to avoid that somehow, but that's how it is now.

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -177,13 +177,12 @@ spec = do
 sqliteDetailedSeqSpec :: Spec
-sqliteDetailedSeqSpec = withDB
-    (fst <$> newFileDBLayer @(SeqState DummyTarget) @DummyTarget @SeqKey) $ do
+sqliteDetailedSeqSpec = withDB newDBSeq $ do
```
@rvl unit tests will fail with this change.  I have used fileBased approach as otherwise in in-memory solution I cannot use SQL side by side. newDBSeq relies on in-memory solution, and to my best knowledge you cannot use runSqlite' with it - it just instantiates next instance of db with no schema and we will end up with : 
```
est/unit/Cardano/Wallet/DB/SqliteSpec.hs:212:9: 
  1) Cardano.Wallet.DB.Sqlite, Sqlite multiple-checkpoints test (SeqState), put and read checkpoints, previous checkpoints are also stored
       uncaught exception: SqliteException
       SQLite3 returned ErrorError while attempting to perform prepare "SELECT \"wallet_id\", \"slot\", \"parent_block_id\", \"block_height\" FROM \"checkpoint\"": no such table: checkpoint
```
Of course, it would be much better if the tests only used DBLayer functions. But....I needed them to make sure every step is reflected in DB tables as expected. I wanted to follow the life-cycle of utxo row, make sure they are not superfluous rows, read spent_utxo column, ... those things are digested by DBlayer methods. At this moment only inspecting DBlayer methods side by side with SQL SELECT I am confident I can follow what exactly happens in DB - hence the tests. But I agree, using DBlayer methods would considerably simplify tests (we could use in-memory solution for example)

paweljakubas replied:
now unit tests should pass in new commit https://github.com/input-output-hk/cardano-wallet/pull/686/commits/380fd075f17a83dbca25d89f5f3048edceb97858

rvl replied:
Oops, I didn't see that it was a different `newDB` function.
There is:
```
unsafeRunQuery :: SqliteContext -> SqlPersistT IO a -> IO a
```
Exported from the Sqlite module and you could add a test function:
```
withDBCtx :: SpecWith (SqliteContext, DBLayer IO s t k) -> Spec
withDBCtx = beforeAll newMemoryDBLayer . beforeWith (\(ctx, db) -> cleanDB db $> (ctx, db))
```
That should allow you to run arbitrary queries, on a memory database, with migrations and logging.

paweljakubas replied:
done. thanks to that both RndState and SeqState was embraced and most dependencies added were unneeded

paweljakubas committed:
make unit tests pass again

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -686,7 +704,7 @@ deleteTxMetas wid = deleteWhere [ TxMetaWalletId ==. wid ]
 putTxMetas :: [TxMeta] -> SqlPersistT IO ()
 putTxMetas metas = dbChunked repsertMany
-    [(TxMetaKey txMetaTxId txMetaWalletId, m) | m@TxMeta{..} <- metas]
+    [(TxMetaKey txMetaTxId txMetaWalletId txMetaSlotId, m) | m@TxMeta{..} <- metas]
```
This looks good.
There also needs to be a change to `selectTxHistory` (in this PR or the next) so that Pending txs are filtered out if there is a subsequent InChain tx.

paweljakubas replied:
changed `selectTxHistory`

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -752,12 +770,68 @@ selectLatestCheckpoint wid = fmap entityVal <$>
+                        , UtxoInputId ==. inpId
+                        , UtxoInputIndex ==. inpIx
+                        ] [LimitTo 1]
+    mapM_ _updateUtxoSpent utxos
```
This should be using `updateWhere`. My model is to try to think of what SQL query would achieve the result, then translate it to persistent.

paweljakubas replied:
corrected. used `updateWhere`

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -870,6 +947,16 @@ instance W.KeyToAddress t Seq.SeqKey => PersistState (Seq.SeqState t) where
+    deleteCheckpointState wid sid = do
+        cpid <- selectList [ SeqStateCheckpointWalletId ==. wid
+                           , SeqStateCheckpointSlot ==. sid ] []
+        deleteWhere [ SeqStatePendingIxCheckpointId <-. fmap entityKey cpid ]
```
Just a note that I made a mistake in the schema design, and these pending indices should not be attached to a checkpoint. But we can leave it as is now and fix the schema and query before doing the rollback function.

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -677,6 +684,17 @@ deleteCheckpoints wid = do
     deleteWhere [CheckpointWalletId ==. wid]
     deleteState @s wid -- clear state
 
+-- | Delete the checkpoint associated with a wallet and slotId.
```
Better mention in the comment that this one is introduced to have guarantee we can put the checkpoints with the same id more than 1 time. Mainly so that the tests can do that and not crash. Otherwise people might wonder why it's needed.

paweljakubas replied:
done

rvl reviewed (commented):
This seems to implement the AC and functions specified in the issue, except for readTxHistory.
The utxo update function could be reworked to use a relational query.
Property tests that use only DBLayer functions will come in #724. We need to improve the generators a bit. I really hope that we can limit the number of tests that need to run arbitrary sqlite queries.

paweljakubas committed:
replace runSqlite' with unsafeRunQuery and add parametrize for both RndState and SeqState

remove unneeded dependencies

correct dependencies

nix patch

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -752,12 +770,68 @@ selectLatestCheckpoint wid = fmap entityVal <$>
+                        , UtxoOutputAddress ==. outAddr
+                        , UtxoOutputCoin ==. outAmt
+                        ] []
+    when (null utxos) $ insert_ utxo
```
can be removed as hinted before, unless I am missing something :thinking: ?

paweljakubas replied:
well, I want to insert only new utxo, if the utxo was present in previous slots I do not want to do anything

KtorZ replied:
Leave that to the database: `INSERT INTO utxos <values> WHERE NOT EXISTS (SELECT * FROM utxos WHERE id IN <utxos_ids>)` ;)

rvl replied:
Unfortunately I don't think that we can express subqueries with plain persistent functions. Anyway I have removed this function.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -752,12 +770,68 @@ selectLatestCheckpoint wid = fmap entityVal <$>
+    -> SqlPersistT IO ()
+updateUtxosPerMeta txins (TxMeta txid wid _ _ sl _) = do
+    let validTxIns = filter (\(TxIn txid' _ _ _ _ ) -> txid'==txid) txins
+    mapM_ (updateUtxoSpent wid sl) validTxIns
```
We can be much simpler about these two functions above:

- For each transaction input, update if exists, all UTxO from our wallet for matching input id and input ix.

A UTxO is uniquely identified by its input id and input index. So there are two scenarios: 

- Either the UTxO is known and in the database, in which case, we have spent it.
- We don't know this (inputId, inputIx) and the update is without any effect

We should avoid putting some business logic inside the database layer. There's no interpretation of the data happening, no need to check for direction, no need to verify transaction ids :wink:

rvl replied:
Done

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -815,6 +889,9 @@ class PersistState s where
     -- | Remove the state for all checkpoints of a wallet.
     deleteState :: W.WalletId -> SqlPersistT IO ()
+    -- | Remove the state for one checkpoint of a wallet.
+    deleteCheckpointState :: W.WalletId -> W.SlotId -> SqlPersistT IO ()
```
There seems to be some duplication between `deleteCheckpoints` and `deleteCheckpoint` which ripples down to here and `deleteState`.  We only use `deleteCheckpoints` and `deleteState` when removing an entire wallet.

What about simply allowing an extra parameter to deleteState and deleteCheckpoints to support passing one or several slot ids:

```hs
-- | Delete the checkpoint associated with a wallet and slotId.
deleteCheckpoints
    :: forall s. PersistState s
    => W.WalletId
    -> [W.SlotId]
    -> SqlPersistT IO ()
deleteCheckpoints wid sids = do
    deleteWhere [UtxoWalletId ==. wid, UtxoSlot <-. sids]
    deleteWhere [CheckpointWalletId ==. wid, CheckpointSlot <-. sids]
    deleteState @s wid sids

deleteState :: W.WalletId -> [W.SlotId] -> SqlPersistT IO ()
deleteState wid sids = do
    cpid <- selectList
        [ SeqStateCheckpointWalletId ==. wid
        , SeqStateCheckpointSlot <-. sids
        ] []
    deleteWhere
        [ SeqStatePendingIxCheckpointId <-. fmap entityKey cpid
        ]
    deleteCascadeWhere
        [ SeqStateCheckpointWalletId ==. wid
        , SeqStateCheckpointSlot <-. sids
        ]
    deleteWhere
        [ SeqStateAddressWalletId ==. wid
        , SeqStateAddressSlot <-. sids
        ]
```

Then, when deleting a wallet, we simply need to first query all the slot ids of all checkpoints, and make sure to pass them as parameter. It's not the most efficient, but for removing wallet which is a fairly rare operation, it sounds pretty okay IMO.

rvl replied:
I changed delete checkpoints to take Maybe SlotId because It's either one checkpoint or all that we want to delete.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -870,6 +947,16 @@ instance W.KeyToAddress t Seq.SeqKey => PersistState (Seq.SeqState t) where
+        deleteCascadeWhere [ SeqStateCheckpointWalletId ==. wid
+                           , SeqStateCheckpointSlot ==. sid ]
+        deleteWhere [ SeqStateAddressWalletId ==. wid
+                    , SeqStateAddressSlot ==. sid ]
```
Small note on indentation here:

```hs
       deleteWhere [ SeqStateAddressWalletId ==. wid
                    , SeqStateAddressSlot ==. sid ]
```

vs

```hs
       deleteWhere 
            [ SeqStateAddressWalletId ==. wid
            , SeqStateAddressSlot ==. sid 
            ]
```

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs:
```diff
@@ -153,12 +193,290 @@ sqliteSpec = withDB (fst <$> newMemoryDBLayer) $ do
+    -> SpecWith (SqliteContext, DBLayer IO s DummyTarget k)
+multipleCheckpointsSpec initialState = do
+    describe "put and read checkpoints" $ do
+        it "previous checkpoints are also stored" $ \(ctx,db) -> do
```
Hmmm... That's quite complicated :thinking: 
It'd be nice to outline what we're trying to "prove" here. If I understand correctly:

- We want to verify that we can store new checkpoints and preserve past ones. 
- I don't understand what others scenarios are trying to achieve :cry: (sorry)

To me, it seems like we have gone through a wrong path here and we need to step back a little. Also, I don't understand why we query the database directly instead of using the db layer. We build the db layer to construct also an invariant-preserving interface. We shouldn't have to use anything else; that's a path which could be quite error-prone. So let's stick with the DBLayer functions. We aren't testing against an implementation, we are testing against an interface!

Also, it seems that all tests contains a lot of arbitrary data ? Why not generating transactions and checkpoints (we already have code for that in the `DBSpec` module!).  We could come up with properties in the very same way we've done it before. For example: calling putCheckpoint `n` times, and `readcheckpoint` should give us the last entry; actually we already have such property :smile: (cf `prop_sequentialPut` !) 

I think we have to start over with testing here and, begin with stating what we want to test. Then we can think about what's the most suitable way of testing it (unit, property, state-machine, integration ...)

paweljakubas committed:
address part of PR review

compilation fix

rvl committed:
Don't use SqliteContext in tests

Also make the tests slightly more debuggable.

rvl committed:
Combine deleteCheckpoint and deleteCheckpoints

rvl committed:
Change mkTxHistory so that it's easier to mark utxo spent

rvl committed:
insertMany utxo -> putMany utxo

rvl committed:
Hlint

rvl committed:
Fix the Any insertCheckpointState

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Model.hs:
```diff
@@ -162,7 +162,7 @@ mListWallets db@(Database wallets _) = (Right (sort $ Map.keys wallets), db)
 mPutCheckpoint :: Ord wid => wid -> Wallet s t -> ModelOp wid s t xprv ()
 mPutCheckpoint wid cp = alterModel wid $ \wal ->
-    ((), wal { checkpoints = [cp] })
+    ((), wal { checkpoints = (cp:checkpoints wal) })
```
I love how this change is one word in the basic model and 100 lines in the SQLite implementation :sweat_smile:

KtorZ committed:
review type-signature for 'setUtxoSpent' and add comment

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -752,12 +772,23 @@ selectLatestCheckpoint wid = fmap entityVal <$>
+        , UtxoSlotSpent ==. Nothing
         ] []
 
+-- fixme: this definitely needs an explanation
```
Done :+1:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -310,9 +313,12 @@ newDBLayer logConfig trace fp = do
+                      let (BlockHeader sid _) = currentTip cp
+                      deleteCheckpoints @s wid (Just sid)
+                      deleteLooseTransactions
+                      insertCheckpoint wid cp
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -355,13 +361,15 @@ newDBLayer logConfig trace fp = do
+                      let (metas, txins, txouts) = flatTxHistory entities
                       putTxMetas metas
                       putTxs txins txouts
+                      softDeleteUTxO wid (zip (txMetaSlotId <$> metas) txins)
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -564,21 +573,25 @@ mkTxHistory
+flatTxHistory entities =
+    ( map fst entities
+    , concatMap (fst . snd) entities
+    , concatMap (snd. snd) entities )
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -564,21 +573,25 @@ mkTxHistory
-    , concatMap (dist mkTxOut . ordered (W.outputs @t)) txs )
+mkTxInputsOutputs tx =
+    ( (dist mkTxIn . ordered (resolvedInputs @t)) tx
+    , (dist mkTxOut . ordered (W.outputs @t)) tx )
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -664,18 +677,25 @@ insertCheckpoint wid cp = do
     putTxMetas metas
     putTxs ins outs
-    dbChunked insertMany_ utxo
+    dbChunked putMany utxo
```
The primary key for `UTxO` is defined as:

```
    Primary
        utxoWalletId
        utxoSlot
        utxoInputId
        utxoInputIndex
        utxoOutputAddress
        utxoOutputCoin
```

So, if we have a different `slot`, we have a different UTxO / row from SQLite's perspective. So here, we are effectively re-writing many times the same UTxO  but with different slot id.

KtorZ replied:
What if we remove _"superfluous"_ constraints from the primary key? A UTxO should actually be uniquely identified by:

- A Wallet Id
- An Input Id
- An Input Index

KtorZ committed:
remove slot, output address and output coin from UTxO primary key

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -664,18 +677,25 @@ insertCheckpoint wid cp = do
+    deleteWhere $
+        (CheckpointWalletId ==. wid) :
+        [CheckpointSlot ==. sid | Just sid <- [mSlot]]
+    deleteState @s wid mSlot -- clear state
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -752,12 +772,27 @@ selectLatestCheckpoint wid = fmap entityVal <$>
         [ UtxoWalletId ==. wid
-        , UtxoSlot ==. sl
+        , UtxoSlotSpent ==. Nothing
         ] []
```
:+1:

KtorZ committed:
refactor select tx history pending pruning

It was quite unclear what we were doing by reading
the code. So I've factored it out in a named function,
trying to better convey the intention.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -863,12 +911,17 @@ instance W.KeyToAddress t Seq.SeqKey => PersistState (Seq.SeqState t) where
+            (SeqStateAddressWalletId ==. wid) :
+            [SeqStateAddressSlot ==. sid | Just sid <- [mSlot]]
+        when (isNothing mSlot) $
+            deleteWhere [SeqStateWalletId ==. wid]
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -779,7 +824,8 @@ selectTxHistory wid order conditions = do
     let txids = map txMetaTxId metas
     (ins, outs) <- selectTxs txids
-    pure $ txHistoryFromEntity @t metas ins outs
+    return $ prunePendingDuplicate $ txHistoryFromEntity @t metas ins outs
```
It's a bit sad that our design forces us to traverse the tx history three times in order to remove duplicate pending transactions that happened to also be in ledger.
I'll try to work around that in a next PR...

KtorZ committed:
rename 'createTransaction' to 'postTransaction' in Api.Server for consistency

KtorZ committed:
replace pending transactions with discovered transactions

The idea of having to store two types of the same transaction, one
pending and one in ledger. Especially because we have to filter the
duplicates out by traversing I believe we can handle rollbacks nicely by
keeping a simple row. Here's the idea:

• When transactions are created, we mark them as pending and insert
   them in the DB, with a slot_id equals to the moment they got created (no
   change compared to now)

• When transactions are discovered, we replace / overwrite the previous
   transactions with the new ones, which update their state to "InLedger"
   and the slot_id at which the transaction was discovered. Incoming
   transactions are simply inserted.

• Now, in the case of rollback:

   • We discard every Incoming transaction beyond the point of
      rollback, they were not created by us and we haven't discovered them.

   • We update every Outgoing transaction beyond the point of rollback
      in such a way that their slot_id becomes the point of rollback, and
      their status become Pending again.

:information_source: This also covers the case of transactions going
from the wallet to itself, because they are marked as Outgoing; our
semantic for these directions is:

• Outgoing The wallet balance decreases.
• Incoming The wallet balance increases or stays the same.

Conceptually, it means that we interpret slot_id as the last moment the
transaction has changed status. There's no pruning needed, because
there's no duplicated rows. When we rollback, we effectively forget the
discovered transaction, and re-insert a pending transaction so the
approach is also quite intuitive IMO.

KtorZ committed:
Fix missing unique constraint on UTxO fields (cf comment about 'upsert')

KtorZ committed:
review grouping of txin with metadata & UTxO soft deletion
